test.sh

```diff
@@ -60,6 +60,78 @@ fi
 
 AMI_ID=$(aws ec2 describe-images --region $region --filters "Name=name,Values=TaupageBuild-$TAUPAGE_VERSION" --query 'Images[*].{ID:ImageId}' --output  text)
 
+
+echo "Starting gpu server..."
+result=$(aws ec2 run-instances \
+    --image-id $AMI_ID \
+    --count 1 \
+    --instance-type p2.xlarge \
+    --instance-initiated-shutdown-behavior terminate\
+    --associate-public-ip-address \
+    --key-name $keypair \
+    --security-group-ids $security_group \
+    --subnet-id $subnet \
+    --output json \
+    --region $region \
+    --placement "AvailabilityZone=eu-central-1a"
+    --tag-specifications 'ResourceType=instance,Tags=[{Key=Taupage,Value=TestMachine}]' \
+    --user-data file://$(pwd)/test-gpudata.yaml)
+
+instanceid=$(echo $result | jq .Instances\[0\].InstanceId | sed 's/"//g')
+echo "Instance: $instanceid"
+aws ec2 create-tags --region $region --resources $instanceid --tags "Key=Name,Value=Taupage AMI GPU Test, Key=Version,Value=$TAUPAGE_VERSION"
+
+while [ true ]; do
+    result=$(aws ec2 describe-instances --region $region --instance-id $instanceid --output json)
+    ip=$(echo $result | jq .Reservations\[0\].Instances\[0\].PublicIpAddress | sed 's/"//g')
+
+    [ ! -z "$ip" ] && [ "$ip" != "null" ] && break
+
+    echo "Waiting for public IP..."
+    sleep 5
+done
+
+echo "IP: $ip"
+
+# wait for server
+while [ true ]; do
+    echo "Waiting for server..."
+
+    set +e
+    ssh $ssh_args ubuntu@$ip echo >/dev/null
+    alive=$?
+    set -e
+
+    if [ $alive -eq 0 ]; then
+        break
+    fi
+
+    sleep 2
+done
+
+terminationcounter=300
+
+while [ true ]; do
```

Comment:
why not a for loop here?
`for i in {1..300}; do ...; done`

Reply:
Because I just copied a similar test and made some small changes.

Reply:
Also with a fixed number of attempts there will have to be error handling and in bash. This way the pipeline just times out. Also it's expected that this step won't fail.

Reply:
true, but with a for loop you do not need to use the bash math expression and reduce the lines of code.